text/0069-api-v2.md

```diff
@@ -0,0 +1,178 @@
+3. Ensure that no non-TiDB client is online.
+4. Use `pd-ctl` to enable `API V2`.
+5. Enable `API V2` in TiKV config file and restart TiKV.
+
+### Downgrade
+
+Downgrade from `API V2` to `API V1` is also simple:
+
+1. Ensure that no non-TiDB TiKV client is online.
+2. Ensure that no non-TiDB data is in TiKV. Detele if any.
+3. Use `pd-ctl` to disable `API V2`.
+4. Disable `API V2` in TiKV config file and restart TiKV.
+
+### CDC / BR
+
+Since all access to TiDB is unchanged during the upgrade, CDC and BR should work trivally on TiDB data.
+
+### Data migration
+
+It's reasonable to provide a way to import and export non-TiDB data in TiKV during the upgrade or downgrade. On TiKV before 4.0, the only way to do that is `scan` and `batch_put` on the client. After 4.0, TiKV start to support importing SST file into TxnKV, and after 5.1, importing on RawKV is also supported. You can find more information in [`RFC: Online Bulk Load for RawKV`](https://github.com/tikv/rfcs/pull/72).
+
+### Client Version
+
+TiKV clients will have a new configuration choosing `API V1` or `API V2` mode on connecting to TiKV.
+
+TiKV will reject the request if the client mode is mismatched to the TiKV configuration. There is one exception: With TiKV in `API V2` mode, you can use `TxnClient` in `API V1` mode to access the TiDB data (key starting with `m` or `t`) which is invisible to the client in `API V2` mode.
+
+Note that in `API V2` mode, you can't specify CF in `RawClient`.
+
+Using TxnKV along with RawKV is safe in `API V2` mode.
+
+Listed below is the detailed compatibility matrix:
+
+|              | V1 Server | V2 Server |
+| ------------ | --------- | --------- |
+| V1 RawClient | Raw Data  | Forbidden |
+| V1 TxnClient | Txn Data  | TiDB Data |
+| V2 RawClient | Forbidden | Raw Data  |
+| V2 TxnClient | Forbidden | Txn Data  |
+
+### Keyspace
+
+When API V2 is enabled for the client, the user will be able to specify a keyspace for a session of `RawClient` or `TxnCient`. A keyspace is identified by a string, whose default value is 'default'.
+
+Keys in other keyspace will be invisible to the user.
+
+## Reference-level design
+
+### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x{keyspace prefix id}`: TxnKV key.
+3. `r{keyspace prefix id}`: RawKV key.
+4. `a{keyspace prefix id}`: RawKV atomic key, on which you can use `CompareAndSwap`.
+
+The `{keyspace prefix id}` is the [keyspace](https://github.com/tikv/rfcs/pull/39) prefix for seperating keys of different keyspace. It should be an vary-length integer whose highest bit of every byte denotes whether the next byte is still part of the integer. The client will fetch the prefix from PD by the keyspace name specified by the user when initializing the client, so it means that the keyspace prefix is valid during the session, in other words, the change on keyspace on PD will not take effect on running seesions.
```

Comment:
how to utilize key space in the tidb side?

Reply:
TiDB doesn't have a corresponding keyspace because it uses the API V1 reserving for TiDB.